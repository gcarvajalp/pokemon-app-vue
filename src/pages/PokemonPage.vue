<template>
  <h1 v-if="!pokemon">Espere por favor ...</h1>

  <div v-else>
    <h1>¿Quien es este pokemón?</h1>
    <PokemonPicture :pokemonId="pokemon.id" :showPokemon="showPokemon" />
    <PokemonOptions :pokemons="pokemonArr" @selection="checkAnswer" />

    <template v-if="showAnswer">
      <h2 class="fade-in">{{ message }}</h2>
      <button @click="newGame">Nuevo Juego</button>
    </template>
  </div>
</template>

<script>
import PokemonOptions from "@/components/PokemonOptions";
import PokemonPicture from "@/components/PokemonPicture";

import getPokemonOptions from "@/helpers/getPokemonOptions";

// console.log(getPokemonOptions())

export default {
  name: "PokemonPage",
  data() {
    return {
      pokemonArr: [],
      pokemon: null,
      showPokemon: false,
      showAnswer: false,
      message: "",
    };
  },
  components: {
    PokemonOptions,
    PokemonPicture,
  },
  methods: {
    async mixPokemonArray() {
      this.pokemonArr = await getPokemonOptions();
      const rndInt = Math.floor(Math.random() * 4);
      this.pokemon = this.pokemonArr[rndInt];
    },
    checkAnswer(pokemonId) {
      this.showPokemon = true
      this.showAnswer = true
      this.message =
        this.pokemon.id === pokemonId
          ? `Muy bien!, es ${this.pokemon.name}`
          : `Oh! es ${this.pokemon.name}, sigue practicando!`;
    },
    newGame(){
      this.showPokemon = false
      this.showAnswer = false
      this.pokemon = null
      this.message = ''
      this.pokemonArr = []
      this.mixPokemonArray();  
    }
  },
  mounted() {
    this.mixPokemonArray();
  },
};
</script>

<style>
</style>