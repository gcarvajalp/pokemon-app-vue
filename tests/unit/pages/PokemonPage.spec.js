import { shallowMount, mount } from '@vue/test-utils'
import PokemonPage from '@/pages/PokemonPage'
import { pokemons } from '../mocks/pokemons.mock'


describe('PokemonPage Component', () => {

  let wrapper

  beforeEach(() => {
    wrapper = shallowMount(PokemonPage)
  })

  test('snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  test('debe llamar mixPokemonArray al montar', () => {

    const mixPokemonArraySpy = jest.spyOn(PokemonPage.methods, 'mixPokemonArray')
    const wrapper = shallowMount(PokemonPage)
    expect(mixPokemonArraySpy).toHaveBeenCalled()

  })

  test('debe hacer match con el snapshot cuando cargan los pokemon', () => {

    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[1],
          showPokemon: false,
          showAnswer: false,
          message: "",
        }
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  test('debe mostrar los componentes de PokemonPicture y PokemonOptions', () => {

    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[1],
          showPokemon: false,
          showAnswer: false,
          message: "",
        }
      }
    })

    expect(wrapper.find('pokemon-picture-stub').exists()).toBeTruthy()
    expect(wrapper.find('pokemon-options-stub').exists()).toBeTruthy()
    expect(wrapper.find('pokemon-picture-stub').attributes('pokemonid')).toBe('2')
    expect(wrapper.find('pokemon-options-stub').exists('pokemons')).toBe(true)
    // console.log(wrapper.find('pokemon-options-stub').attributes())

  })

  test('pruebas con checkAnswer', async () => {

    const wrapper = shallowMount(PokemonPage, {
      data() {
        return {
          pokemonArr: pokemons,
          pokemon: pokemons[0],
          showPokemon: false,
          showAnswer: false,
          message: "",
        }
      }
    })

    await wrapper.vm.checkAnswer(1)

    expect(wrapper.find('h2').exists()).toBe(true)
    expect(wrapper.vm.showPokemon).toBe(true)
    expect(wrapper.find('h2').text()).toBe(`Muy bien!, es ${pokemons[0].name}`)
    
    await wrapper.vm.checkAnswer(3)

    expect(wrapper.find('h2').text()).toBe(`Oh! es ${pokemons[0].name}, sigue practicando!`)

    expect(wrapper.vm.message).toBe(`Oh! es ${pokemons[0].name}, sigue practicando!`)
    
  })





})
