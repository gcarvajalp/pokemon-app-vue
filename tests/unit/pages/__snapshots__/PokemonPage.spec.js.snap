// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonPage Component debe hacer match con el snapshot cuando cargan los pokemon 1`] = `
<div>
  <h1>¿Quien es este pokemón?</h1>
  <pokemon-picture-stub pokemonid="2" showpokemon="false"></pokemon-picture-stub>
  <pokemon-options-stub pokemons="[object Object],[object Object],[object Object],[object Object]"></pokemon-options-stub>
  <!--v-if-->
</div>
`;

exports[`PokemonPage Component snapshot 1`] = `<h1>Espere por favor ...</h1>`;
